File: Nesol.R
# Solubility of Ne in sea water
#=========================================================================
#  2/1/18
#            Original Author: Roberta C. Hamme (Scripps Inst of Oceanography)
# translating/editing author: Joseph D. Zottoli (URI GSO)
#
# USAGE:  eqsolubility(S,t)
#
# DESCRIPTION:
#    Solubility (saturation) of neon (Ne) in sea water
#    at 1-atm pressure of air including saturated water vapor
#
# INPUT:  (if S and t are not singular they must have same dimensions)
#   S = salinity    (PSS)
#   t = temperature (degree C)
#
# OUTPUT:
#   conc_Ne = solubility of Ne  [umol/kg] 
# 
# AUTHOR:  Joseph Zottoli (jzottoli@uri.edu)
#
# REFERENCE:
#    Roberta Hamme and Steve Emerson, 2004.
#    "The solubility of neon, nitrogen and argon in distilled water and seawater."
#    Deep-Sea Research I, 51(11), p. 1517-1528.
#
# DISCLAIMER:
#    This software is provided "as is" without warranty of any kind.  

#start function
eqsolubility=function(S,te){ 
  conc_Ne<<-NA
  if(length(S)!=length(te)) stop("S and te must be of same length")
  temp_S=rep(NA,length(te)) 
  
  for (i in 1:length(te)){
    # convert T to scaled temperature
    
    temp_S[i] = log((298.15 - te[i])/(273.15 + te[i]))
    
    # Eqn (1) of Hamme and Emerson 2004 with constants from table 4 inserted
    conc_Ne[i] = exp(2.18156 + 1.29108*(temp_S[i]) + 2.12504*temp_S[i]^2 + S[i]*(-5.94737e-3 + -5.13896e-3*temp_S[i]))
    
    # Convert from nmol/kg to umol/kg
    conc_Ne[i] <<- conc_Ne[i]/1000
    
  }
  print(conc_Ne)
}

eqsolubility(35,te=10)#shows equllibrium solubility of one temop and salinity

eqsolubility(c(35,33,27,38,30),te=c(10,15,25,20,10))#shows equllibrium solubility of 10 combinations

eqsolubility(c(35,33),te=c(20,10,5))#shows that vectors must be of same length

File: Nesol.py
# -*- coding: utf-8 -*-
"""
    Gas Solubility Package for Python
        Version 2.0 October, 2022
        Author:  Brice Loose, Graduate School of Oceanography, URI.
        Email: bloose@uri.edu

    These noble gas solubility functions are being updated to reflect the new
    solubility determinations for He, Ne, Ar, Kr, and Xe from 
    Jenkins et al., (2019).  These replace the prio relationships developed by
    Roberta Hamme.
    
    #=========================================================================
    # Hesol   Solubility of Ne in sea water
    #=========================================================================
    #
    # USAGE:  concNe = Nesol(S,T)
    #
    # DESCRIPTION:
    #    Solubility (saturation) of neon (Ne) in sea water 
    #    at 1-atm pressure of air including saturated water vapor
    #
    # INPUT:  (if S and T are not singular they must have same dimensions)
    #   S = salinity    [PSS]
    #   T = temperature [degree C]
    #
    # OUTPUT:
    #   concHe = solubility of Ne  [umol/kg] 
    # 
    #
    # REFERENCE:
    #    Jenkins, W., Lott, D.E., Cahill, K.L. (2019) A determination of 
    # atmospheric helium, neon, argon, krypton, and xenon solubility 
    # concentrations in water and seawater.  Marine Chemistry,
    # DOI: https://doi.org/10.1016/j.marchem.2019.03.007
    #
    # DISCLAIMER:
    #    This software is provided "as is" without warranty of any kind.  
    #=========================================================================
"""


def Nesol(S, T):

    import numpy as np

    #----------------------
    # Check input parameters
    #----------------------
    #Python Takes care of this...

    # CHECK S,T dimensions and verify consistent
    if np.size(np.shape(T))>1:
        mt,nt=np.shape(T)
    else:
        mt = np.size(T)
        nt = 0
    if np.size(np.shape(S))>1:
        ms,ns=np.shape(S)
    else:
        ms = np.size(S)
        ns = 0

    # Check that T&S have the same shape or are singular
    if (ms != mt) or (ns != nt):
        print ("Nesol: S & T must have same dimensions or be singular")
        return

    #------
    # BEGIN
    #------

    # convert T to scaled temperature
    temp_abs = np.add(T,273.15)/100

    A1 = -274.1329
    A2 = 352.6201
    A3 = 226.9676
    A4 = -37.13393
    B1 = -0.063860
    B2 = 0.035326
    B3 = -0.0053258
    C1 = 1.28e-5

    # Eqn (2) of Weiss and Kyser
    conc_Ne = np.exp(A1 + (A2 / temp_abs) +
                     (A3 * np.log(temp_abs)) + 
                     (A4 * temp_abs) + S*(B1 +
                    (B2 * temp_abs) + 
                    (B3 * np.power(temp_abs,2))) + C1*S**2)

    
    return conc_Ne*1e6  # OUTPUT IN umol/kg

    